src/app.rs

```diff
@@ -284,12 +287,24 @@ impl Application {
             // Assigned to prevent the API terminating when falling out of scope.
             let api_server = if api_config.enabled {
                 use std::sync::{Arc, atomic::AtomicBool};
-                emit!(ApiStarted {
-                    addr: api_config.address.unwrap(),
-                    playground: api_config.playground
-                });
 
-                Some(api::Server::start(topology.config(), topology.watch(), Arc::<AtomicBool>::clone(&topology.running)))
+                let api_server = api::Server::start(topology.config(), topology.watch(), Arc::<AtomicBool>::clone(&topology.running));
+
+                match api_server {
+                    Ok(api_server) => {
+                        emit!(ApiStarted {
+                            addr: api_config.address.unwrap(),
+                            playground: api_config.playground
+                        });
+
+                        Some(api_server)
+                    }
+                    Err(e) => {
+                        error!("An error occurred that Vector couldn't handle: {}.", e);
+                        let _ = self.config.graceful_crash_sender.send(());
+                        None
+                    }
```

Comment:
I'm not 100% up to speed on whether we want to catch all errors or just some.

The problem statement in #14157 suggests its specific to listening on a socket.

Adding @bruceg to the review to help ensure the scope of the changes is correct.

Reply:
Sure. If we want to distinguish different errors here (however, I do not think that we need to do it here), it could be simply fixed with an additional error code check.

Reply:
@bruceg what do you think about this change?

Reply:
I think this is an improvement in the current situation. My testing indicates that at least some other components (I tested a `http_server` source) simply crash with a panic.